daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala

```diff
@@ -22,15 +24,40 @@ class DarReader[A](
 
   def readArchive(darFile: ZipFile): Try[Dar[A]] =
     for {
-      names <- parseDalfNamesFromManifest(darFile): Try[Dar[String]]
+      names <- readDalfNames(darFile): Try[Dar[String]]
       main <- parseOne(darFile)(names.main): Try[A]
       deps <- parseAll(darFile)(names.dependencies): Try[List[A]]
     } yield Dar(main, deps)
 
+  private def readDalfNames(darFile: ZipFile): Try[Dar[String]] =
+    parseDalfNamesFromManifest(darFile).recoverWith {
+      case NonFatal(e1) =>
+        findDalfNames(darFile).recoverWith {
+          case NonFatal(e2) => Failure(InvalidDar(e1, e2))
+        }
+    }
+
   private def parseDalfNamesFromManifest(darFile: ZipFile): Try[Dar[String]] =
     bracket(Try(darFile.getInputStream(manifestEntry)))(close)
       .flatMap(is => readDalfNamesFromManifest(is))
 
+  private def findDalfNames(darFile: ZipFile): Try[Dar[String]] = {
+    val entries: List[ZipEntry] = darFile.entries.asScala.toList
+    val dalfs: List[String] = entries.filter(isDalf).map(_.getName)
+    dalfs.partition(isPrimDalf) match {
```

Comment:
This won't work for DARs that also contain the stdlib, which is now produced as a separate DALF file next to `daml-prim.dalf` and the DALF file for the actual model. Unless we never want to support a DAR file with no manifest but more than 2 DALF files.

Reply:
> Unless we never want to support a DAR file with no manifest but more than 2 DALF files.
that is my assumption here. @robin-da can you comment?

Reply:
@francesco-da legacy DAR only contained main and prim dalfs, no stdlib. But again, specification for the DAR format would be nice to have

Reply:
```
leo   [42 minutes ago]
Yep, that's what I remember. The old dar that didn't have Main-Dalf in the manifest could only have two dalfs, main and prim. So no reason to overcomplicate the dar reader. New dars with stdlib will always have Main-Dalf in the manifest. Am I right? @robin
And yeah we need a spec, if format keeps changing (edited)


robin   [36 minutes ago]
yes
```